### PR TITLE
Added Unroll All Checkbox for Workspace History Window

### LINKED ISSUE #27033: Add an unroll all button to WS History
Currently users need to click every single workflow algorithm they want to unroll in the history view.

Instead we should have a checkbox or button which unrolls all of the workflow algorithms in the history for them

For some sample data with multiple workflow algs:
- Download ISIS Sample Data
- Extract the folder and open Mantid
- Open the ISIS Sans interface
- Hit user file and open `mask.txt`
- Hit batch file and open `batch.csv`
- Click a row and hit process selected
- Save your workspace somewhere to avoid this hassle again

### PR DESCRIPTION
**Pull request by Matt-Cumber**
**Description of work.**
There is a new checkbox "Unroll All" when viewing the Show History Window for a workspace. This checkbox checks and unchecks the unroll column for all parent algorithms in the tree. This saves time for users wanting to unroll all algorithms.

**To test:**

1. Load in the sample workspaces by following instructions from the original [issue](https://github.com/mantidproject/mantid/issues/27033)
2. Show Algorithm History for some of the workspaces and follow the rest of the instructions
3. Check Unroll All from the tree widget and try copy script to clipboard and copy script to file ensuring all child algorithms are copied correctly
4. Try checking unroll for a few algorithms and then unroll all to see if they are all checked/unchecked

Fixes #27033 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/widgets/common/src/AlgorithmHistoryWindow.cpp [hunk 1]
```diff
@@ -509,6 +517,23 @@ void AlgorithmHistoryWindow::doUnroll(const std::vector<int> &unrollIndicies) {
 
 void AlgorithmHistoryWindow::doRoll(int index) { m_view->roll(index); }
 
+void AlgorithmHistoryWindow::unrollAll(int state) {
+  // Iterate all items in tree which have children algorithms to be unrolled
+  QTreeWidgetItemIterator it(m_Historytree,
+                             QTreeWidgetItemIterator::HasChildren);
+
+  auto head = m_Historytree->headerItem();
+  std::cout << head->text(0).toStdString() << std::endl;
```
Comment (martyngigg) [inline qt/widgets/common/src/AlgorithmHistoryWindow.cpp:526]: Could we remove the `cout`? I assume this was intended for debugging?

**Review by martyngigg — CHANGES_REQUESTED** [review]
This is good focused change and the code looks good.

My only concern with implementing it has a third column is that by default the dialog cuts off the column and you have to scroll to see what it does:

Screenshot from 2020-08-10 11-47-30

Also, oddly, when I first go to click the box I have to click twice: once to select the row and then another time to check the box. Hopefully this gif helps demonstrate this:

workspace-history-unroll-all

I was wondering if we should consider putting the checkbox below the tree:

workspace-history-unroll-option1
 
It would avoid the whole row selection problem and not require scrolling horizontally in the tree. What do you think?

File: qt/widgets/common/src/AlgorithmHistoryWindow.cpp (hunk 1 above)
Comment (Matt-Cumber) [reply qt/widgets/common/src/AlgorithmHistoryWindow.cpp:526]: Good spot, I've removed this now. You are right this was for debugging/clarification

**Comment by Matt-Cumber** [comment]
Interesting images/gifs. I guess I'm using a wider monitor as for me the column is visible by default and one click was all that was needed so I missed this. I spoke to @alicerussell1 regarding the placement of the checkbox. Originally it was placed along side the buttons at the bottom of the widget, but Alice and I agreed it might be missed as it wasn't near any of the other unroll checkboxes.

File: qt/widgets/common/src/AlgorithmHistoryWindow.cpp [hunk 2]
```diff
@@ -509,6 +515,22 @@ void AlgorithmHistoryWindow::doUnroll(const std::vector<int> &unrollIndicies) {
 
 void AlgorithmHistoryWindow::doRoll(int index) { m_view->roll(index); }
 
+void AlgorithmHistoryWindow::unrollAll(int state) {
+  // Iterate all items in tree which have child algorithms to be unrolled
+  QTreeWidgetItemIterator it(m_Historytree,
+                             QTreeWidgetItemIterator::HasChildren);
+
+  auto head = m_Historytree->headerItem();
```
Comment (martyngigg) [inline qt/widgets/common/src/AlgorithmHistoryWindow.cpp:523]: I'm suprised the CI didn't warn on this. I have a compiler warning now that the debugging statement has been removed:
``` 
[6/39] Building CXX object qt/widgets/common/CMakeFiles/MantidQtWidgetsCommonQt5.dir/src/AlgorithmHistoryWindow.cpp.o
../../qt/widgets/common/src/AlgorithmHistoryWindow.cpp: In member function ‘void AlgorithmHistoryWindow::unrollAll(int)’:
../../qt/widgets/common/src/AlgorithmHistoryWindow.cpp:523:8: warning: unused variable ‘head’ [-Wunused-variable]
   auto head = m_Historytree->headerItem();
```

**Review by martyngigg — CHANGES_REQUESTED** [review]
Thanks for moving this. I agree that most of the time it's better to keep related things as close together as possible but this means the tree can just contain information about each history step and the checkbox is fairly hard to miss.

There's just the compiler warning to fix now.

**Review by martyngigg — APPROVED** [review]
This looks good now. Nice work :+1:

**Pull request merged**